File: services/backend/src/main/java/com/corely/corely_backend/entity/Product.java
package com.corely.corely_backend.entity;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

import java.math.BigDecimal;
import java.util.List;
import java.util.Map;
import java.util.UUID;

@Entity
@Table(name = "products")
@Getter
@Setter
public class Product extends BaseEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    private String name;

    @Column(unique = true)
    private String slug;

    @Column(unique = true)
    private String sku;

    private BigDecimal price;

    @Column(name = "discount_price")
    private BigDecimal discountPrice;

    @Column(name = "stock_quantity")
    private Integer stockQuantity;

    @Column(columnDefinition = "TEXT")
    private String description;

    // SỬ DỤNG CÁCH NÀY CHO HIBERNATE 6 (Không cần thư viện ngoài)
    @JdbcTypeCode(SqlTypes.JSON)
    @Column(columnDefinition = "jsonb")
    private Map<String, Object> specs;

    @Column(name = "thumbnail_url")
    private String thumbnailUrl;

    @Column(name = "is_active")
    private Boolean isActive = true;

    // KHAI BÁO LAZY RÕ RÀNG
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id")
    private Category category;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "brand_id")
    private Brand brand;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "store_id", nullable = false)
    private Store store;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL)
    private List<ProductImage> images;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL)
    private List<ProductVariant> variants;
}

File: services/frontend/corely-app/src/types/product.ts
export interface ProductVariantRequest {
    id?: string;
    name: string;
    sku: string;
    price: number;
    stockQuantity: number;
    imageUrl?: string;
}

export interface ProductCreationRequest {
    name: string;
    sku: string;
    price: number;
    discountPrice?: number;
    stockQuantity: number;
    description?: string;
    specs?: Record<string, unknown>;
    thumbnailUrl?: string;
    storeId: string;
    categoryId: string;
    brandId: string;
    imageUrls?: string[];
    variants?: ProductVariantRequest[];
}

export interface ProductVariantResponse {
    id: string;
    name: string;
    sku: string;
    price: number;
    stockQuantity: number;
    imageUrl?: string;
    isActive: boolean;
    createdAt: string;
}

export interface ProductResponse {
    id: string;
    name: string;
    slug: string;
    sku: string;
    price: number;
    discountPrice?: number;
    stockQuantity: number;
    description?: string;
    specs?: Record<string, unknown>;
    thumbnailUrl?: string;
    isActive: boolean;
    storeId: string;
    storeName: string;
    categoryId: string;
    categoryName: string;
    brandId: string;
    brandName: string;
    imageUrls: string[];
    variants: ProductVariantResponse[];
    createdAt: string;
    updatedAt: string;
}

export interface PageProductResponse {
    totalPages: number;
    totalElements: number;
    size: number;
    content: ProductResponse[];
    number: number;
    first: boolean;
    last: boolean;
}
